internal/publish/model/exposure_model_test.go

```diff
@@ -1360,11 +1420,100 @@ func TestDaysFromSymptomOnset(t *testing.T) {
 	}
 }
 
+func TestReviseKeys_FromFederation(t *testing.T) {
+	t.Parallel()
+
+	cases := []struct {
+		name     string
+		existing *Exposure
+		incoming *Exposure
+		err      error
+	}{
+		{
+			name: "export_import_mismatch",
+			existing: &Exposure{
+				ExposureKey:     []byte{0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15},
+				LocalProvenance: false,
+				ExportImportID:  proto.Int64(2),
+			},
+			incoming: &Exposure{
+				ExposureKey:    []byte{0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15},
+				ReportType:     "fake",
+				ExportImportID: proto.Int64(4),
+			},
+			err: ErrorNotSameFederationSource,
+		},
+		{
+			name: "export_import_same",
+			existing: &Exposure{
+				ExposureKey:     []byte{0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15},
+				LocalProvenance: false,
+				ReportType:      "likely",
+				ExportImportID:  proto.Int64(2),
+			},
+			incoming: &Exposure{
+				ExposureKey:    []byte{0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15},
+				ReportType:     "confirmed",
+				ExportImportID: proto.Int64(2),
+			},
+		},
+		{
+			name: "federation_mismatch",
+			existing: &Exposure{
+				ExposureKey:       []byte{0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15},
+				LocalProvenance:   false,
+				FederationQueryID: "foo",
+			},
+			incoming: &Exposure{
+				ExposureKey: []byte{0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15},
+				ReportType:  "fake",
+			},
+			err: ErrorNotSameFederationSource,
+		},
+		{
+			name: "federation_mismatch",
+			existing: &Exposure{
+				ExposureKey:       []byte{0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15},
+				LocalProvenance:   false,
+				ReportType:        "likely",
+				FederationQueryID: "foo",
+			},
+			incoming: &Exposure{
+				ExposureKey:       []byte{0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15},
+				ReportType:        "confirmed",
+				FederationQueryID: "foo",
+			},
+		},
+	}
+
+	for _, tc := range cases {
+		t.Run(tc.name, func(t *testing.T) {
+			ctx := context.Background()
+			existing := make(map[string]*Exposure)
+			existing[tc.existing.ExposureKeyBase64()] = tc.existing
+
+			incoming := make([]*Exposure, 1)
+			incoming[0] = tc.incoming
+
+			_, err := ReviseKeys(ctx, existing, incoming)
+			if err == nil && tc.err != nil {
+				t.Errorf("missing expected error: %v", tc.err)
+			} else if err != nil && tc.err == nil {
+				t.Errorf("unexpected error: %v", err)
+			} else if !errors.Is(err, tc.err) {
+				t.Errorf("wrong error, want: %v got: %v", tc.err, err)
+			}
```

Comment:
Might be clearer to avoid the 3 sets of checks at the same level and split it instead into `tc.err != nil` and `tc.err == nil` respectively, so perhaps to do this:
```go
if tc.err != nil {
     if !errors.Is(err, tc.err) || err == nil {
          t.Fatalf("wrong error, got: %v want: %v", err, tc.err)
     }
} else if err != nil {
     t.Fatalf("unexpectedly returned an error: %v", err)
}
```